digitalocean/vpcpeering/datasource_vpcpeering.go

```diff
@@ -0,0 +1,139 @@
+package vpcpeering
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/digitalocean/godo"
+	"github.com/digitalocean/terraform-provider-digitalocean/digitalocean/config"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func DataSourceDigitalOceanVPCPeering() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceDigitalOceanVPCPeeringRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				Description:  "The ID of the VPC Peering",
+				ValidateFunc: validation.NoZeroValues,
+				ExactlyOneOf: []string{"id", "name"},
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				Description:  "The name of the VPC Peering",
+				ValidateFunc: validation.NoZeroValues,
+				ExactlyOneOf: []string{"id", "name"},
+			},
+			"vpc_ids": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Computed:    true,
+				Description: "The list of VPCs to be peered",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set: schema.HashString,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceDigitalOceanVPCPeeringRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*config.CombinedConfig).GodoClient()
+	var foundVPCPeering *godo.VPCPeering
+
+	if id, ok := d.GetOk("id"); ok {
+		vpcPeering, _, err := client.VPCs.GetVPCPeering(context.Background(), id.(string))
+		if err != nil {
+			return diag.Errorf("Error retrieving VPC Peering: %s", err)
+		}
+
+		foundVPCPeering = vpcPeering
+	} else if name, ok := d.GetOk("name"); ok {
+		vpcPeerings, err := listVPCPeerings(client)
+		if err != nil {
+			return diag.Errorf("Error retrieving VPC Peering: %s", err)
+		}
+
+		vpcPeering, err := findVPCPeeringByName(vpcPeerings, name.(string))
+		if err != nil {
+			return diag.Errorf("Error retrieving VPC Peering: %s", err)
+		}
+
+		foundVPCPeering = vpcPeering
+	}
+
+	if foundVPCPeering == nil {
+		return diag.Errorf("Error retrieving VPC Peering: %s", fmt.Errorf("vpc peering not found"))
```

Comment:
we could be more precise with the error here. I think you will reach this branch if neither "name" nor "id" was given, for that case, it's not a 404 but more like a bad request (401?) .